### PR TITLE
User namespaces issue 898

### PR DESCRIPTION
**Pull request by xin-cai**
docker user namespaces issue 898

### CODE REVIEW

**Comment by sjfink** [comment]
why did we revert to aufs?

**Comment by xin-cai** [comment]
probably @domdom82 can answer why revert to aufs ...

**Comment by xin-cai** [comment]
Got green playground build #550

**Comment by markusthoemmes** [comment]
@sjfink there is no revert to `aufs`, the commit message is misleading. It seems like the vagrant env was never elevated to `overlay`, probably due to kernel version in the 14.04 image.

File: tools/macos/tweak-dockermachine.sh [hunk 1]
```diff
@@ -9,6 +9,7 @@ MACHINE_NAME=${1:-whisk}
 # Disable TLS.
 docker-machine ssh $MACHINE_NAME "echo DOCKER_TLS=no |sudo tee -a /var/lib/boot2docker/profile > /dev/null"
 docker-machine ssh $MACHINE_NAME "echo DOCKER_HOST=\'-H tcp://0.0.0.0:4243\' |sudo tee -a /var/lib/boot2docker/profile > /dev/null"
+docker-machine ssh $MACHINE_NAME "echo EXTRA_ARGS=\'--userns-remap=default\' |sudo tee -a /var/lib/boot2docker/profile > /dev/null"
```
Comment (domdom82) [inline tools/macos/tweak-dockermachine.sh]: when docker-machine launches the daemon it reads the boot2docker/profile and appends EXTRA_ARGS to the dockerd parameters. there is no special env var for the userns flag. check `/usr/local/etc/init.d/docker` on a docker-machine VM.

**Comment by domdom82** [comment]
@sjfink @markusthoemmes re aufs, that is correct. the commit message was "revert to auf" because I tried to move to overlay but then realized we can't use it with the 3.13 kernel in vagrant at it was only available from 3.18. I also tried to upgrade to ubuntu 16 but the vagrant image has lots of bugs so I stayed with 14.04 for now.
see https://docs.docker.com/engine/userguide/storagedriver/overlayfs-driver/ for details.

File: ansible/roles/invoker/tasks/deploy.yml [hunk 2]
```diff
@@ -14,30 +14,32 @@
...
-      "PORT": 8080 
-    volumes:
-      - "{{whisk_logs_dir}}/invoker{{play_hosts.index(inventory_hostname)}}:/logs"
-      - "/var/lib/docker/containers/:/containers"
-    ports:
-      - "{{invoker.port + play_hosts.index(inventory_hostname)}}:8080"
-    command: /bin/sh -c "/invoker/bin/invoker {{play_hosts.index(inventory_hostname)}} >> /logs/invoker{{play_hosts.index(inventory_hostname)}}_logs.log 2>&1"
+- name: determine docker root dir
+  shell: "docker info 2>/dev/null | grep 'Docker Root Dir' | awk -F ': ' '{print $2}'"
+  register: docker_root_dir
```
Comment (domdom82) [inline ansible/roles/invoker/tasks/deploy.yml]: good idea. we can omit the version tho to get the latest. we have already pinned the api version implicitly by pinning the docker version.

```
curl host:port/info
```
should yield the same results. (calls /1.24/info on docker 1.12)

**Comment by xin-cai** [comment]
green build without user namespace enabled in Travis docker daemon

**Comment by domdom82** [comment]
@perryibm pls review as this is going to affect the invoker. should not do any harm though.

**Comment by domdom82** [comment]
@xin-cai squash and rebase after implementing feedback. thanks

**Comment by domdom82** [comment]
the following steps are planned to get this in:
- push this to master with only travis having user namespaces enabled
- existing systems (ci, dev envs, prod) do not use this feature yet and will gracefully ignore it (docker ignores the --userns=host flag if no namespaces are configured)
- push blue parts to master (can be done independently)
- once the code is in place, update docker daemons (vagrant, docker-machine, ci, prod). this should automatically pick up the feature.

**Comment by xin-cai** [comment]
Get green PG2 Build #576

**Pull request merged**